### PR TITLE
Provide tensor-friendly methods for `torch` and `jax` through `OMEArrow.to_dlpack`

### LINKED ISSUE #33: Add DLPack export + streaming tensor views to OME-Arrow
## Goal
Make OME-Arrow pixel data (from the OME struct / Parquet / Vortex) consumable by **PyTorch** and **JAX** with minimal copies by exposing a small **tensor view API** that can export **DLPack** capsules for selected slices, tiles, or batches.

---

### 1. Tensor View Abstraction
- [ ] Define a `TensorView` (name flexible) returned from the OME-Arrow struct.
- [ ] Supports selecting pixel blocks by indices:
  - scene/image
  - t, z, c
  - spatial ROI (x, y, w, h) or tile index
- [ ] Exposes metadata:
  - `shape`
  - `dtype`
  - `device`
  - `layout` (e.g., CHW, HWC, TZCHW)
  - `strides` (if applicable)
- [ ] Enforce a **canonical default layout**:
  - 2D: `CHW`
  - 5D: `TZCHW`
- [ ] Allow explicit layout overrides.

---

### 2. DLPack Export
- [ ] Implement `TensorView.to_dlpack(...)`
- [ ] Signature example:
  ```python
  to_dlpack(device="cpu" | "cuda", contiguous=True | False)
  ```
- [ ] If the view is not representable as a single strided tensor:
  - [ ] Raise a clear, actionable error **OR**
  - [ ] Materialize once into a contiguous buffer (preferred: directly on target device).
- [ ] Define and document **ownership/lifetime semantics** for the returned DLPack capsule.

---

### 3. Streaming / Batching
- [ ] Implement an iterator-based API:
  ```python
  iter_dlpack(
      batch_size=...,
      tiles=...,
      shuffle=False,
      prefetch=...
  )
  ```
- [ ] Iterator yields one DLPack capsule per batch or tile.
- [ ] Deterministic behavior when a seed is provided (if shuffling supported).

---

### 4. Optional Framework Shims (Nice-to-have)
- [ ] `to_torch()` thin wrapper using DLPack import utilities.
- [ ] `to_jax()` thin wrapper using DLPack import utilities.
- [ ] Keep DLPack as the core interoperability surface (avoid hard framework coupling).

---

### 1. Round-Trip Correctness (CPU)
- [ ] Create a small synthetic OME-Arrow struct (e.g., `C=2, H=8, W=8`) with known values.
- [ ] Export via `to_dlpack()` and import into:
  - [ ] PyTorch (`torch.utils.dlpack.from_dlpack`)
  - [ ] JAX (`jax.dlpack.from_dlpack`, if available)
- [ ] Assert:
  - Shapes match
  - Dtypes match
  - Values match exactly

---

### 2. Contiguity & Layout
- [ ] Verify default layout (CHW) is correct.
- [ ] Verify requested layouts (e.g., HWC) produce expected permutations.
- [ ] If `contiguous=True`, assert resulting tensor is contiguous in the consumer framework.

---

### 3. Zero-Copy Expectations (CPU)
- [ ] Where feasible, assert buffer sharing (e.g., pointer/storage checks).
- [ ] At minimum, assert `contiguous=False` does **not** materialize when the Arrow view is already compatible.

---

### 4. Error Paths
- [ ] Non-representable Arrow layouts trigger documented behavior (error or materialization).
- [ ] Invalid device requests raise clear, helpful errors.

---

### 5. Iterator Semantics
- [ ] `iter_dlpack(batch_size=...)` yields the correct number of batches.
- [ ] Concatenated batch content matches expected values.
- [ ] If shuffle is supported:
  - [ ] Deterministic output with a fixed seed.

---

## Documentation (Minimal)
- [ ] Short docs section: **“Exporting OME-Arrow pixel data via DLPack”**
- [ ] One minimal example snippet for:
  - PyTorch
  - JAX

---

## Acceptance Criteria
- [ ] PyTorch can consume OME-Arrow pixel slices/tiles via DLPack with correct values and layout.
- [ ] JAX can consume the same API (or tests skip cleanly if JAX is unavailable).
- [ ] Device placement, contiguity, and ownership semantics are explicit and predictable.

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This PR provides tensor-friendly methods for loading images in `torch` or `jax` by using `OMEArrow.to_dlpack`. The goal is to provide zero-copy capabilities in tensor environments through Arrow, avoiding unnecessary conversions where possible.

Closes #33 

## What kind of change(s) are included?

- [ ] Documentation (changes docs or other related content)
- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [x] I have read and followed the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] I have searched for existing content to ensure this is not a duplicate.
- [x] I have performed a self-review of these additions (including spelling, grammar, and related).
- [x] These changes pass all pre-commit checks.
- [x] I have added comments to my code to help provide understanding
- [x] I have added a test which covers the code changes found within this PR
- [x] I have deleted all non-relevant text in this pull request template.

## Release Notes

* **New Features**
  * Added DLPack tensor export functionality for zero-copy PyTorch and JAX integration
  * Added `tensor_view()` method for tensor-like access to pixel data with rich indexing support
  * Added optional dependencies for PyTorch and JAX integration

* **Documentation**
  * Added comprehensive DLPack guide with usage examples and compatibility notes
  * Added examples demonstrating PyTorch and JAX workflows

* **Tests**
  * Added extensive test coverage for DLPack export and tensor operations

### CODE REVIEW

File: src/ome_arrow/tensor.py [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+            self._data_py = self._struct_scalar.as_py()
+        return self._data_py
+
+    def _pixels_meta(self) -> dict[str, Any]:
+        if self._data_py is not None:
+            return self._data_py["pixels_meta"]
+        if self._struct_array is None:
+            raise ValueError("No OME-Arrow scalar available.")
+        pm_arr = self._struct_array.field("pixels_meta")
+        return pm_arr[0].as_py()
```
[the remark this answers is not in the document]
Comment (d33bs) [reply src/ome_arrow/tensor.py]: Thanks - addressed.

File: src/ome_arrow/tensor.py [hunk 2]
```diff
@@ -0,0 +1,761 @@
...
+    def shape(self) -> tuple[int, ...]:
+        """Return the tensor shape for the current layout."""
+
+        return self.to_numpy(contiguous=False).shape
+
+    @property
+    def strides(self) -> tuple[int, ...]:
+        """Return the tensor strides in bytes for the current layout."""
+
+        return self.to_numpy(contiguous=False).strides
```
[the remark this answers is not in the document]
Comment (d33bs) [reply src/ome_arrow/tensor.py]: Thanks addressed.

File: src/ome_arrow/tensor.py [hunk 3]
```diff
@@ -0,0 +1,847 @@
...
+            return bool(self._data_py.get("chunks"))
+        if self._struct_array is not None:
+            chunks_arr = self._struct_array.field("chunks")
+            if len(chunks_arr) == 0:
+                return False
+            return not chunks_arr.is_null().to_pylist()[0]
+        if self._struct_scalar is not None:
+            chunks_scalar = self._struct_scalar["chunks"]
+            return bool(chunks_scalar.is_valid)
+        return False
```
[the remark this answers is not in the document]
Comment (d33bs) [reply src/ome_arrow/tensor.py]: Thanks - I believe these are addressed now.

File: src/ome_arrow/tensor.py [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+    return np.transpose(arr, axes)
+
+
+def _dtype_from_meta(dtype: Any) -> np.dtype:
+    if dtype is None:
+        return np.dtype(np.uint16)
+    try:
+        return np.dtype(dtype)
+    except Exception:
+        return np.dtype(np.uint16)
```
[the remark this answers is not in the document]
Comment (d33bs) [reply src/ome_arrow/tensor.py]: Thanks - addressed.

File: pyproject.toml [hunk 5]
```diff
@@ -120,17 +127,8 @@ lint.per-file-ignores."docs/*" = [ "D100", "D101", "D102", "D103", "D104", "D105
...
-[tool.poe.tasks]
-# tasks for development environments
-test.shell = """
-pytest
-"""
-jupyter.shell = """
-jupyter lab
-"""
+[tool.pytest]
+ini_options.pythonpath = [ "." ]
```
[the remark this answers is not in the document]
Comment (d33bs) [reply pyproject.toml]: Thanks - addressed.

File: src/ome_arrow/tensor.py [hunk 6]
```diff
@@ -0,0 +1,877 @@
...
+            layout: Desired layout string using TZCHW letters.
+            dtype: Output dtype override.
+        """
+        self._struct_array: pa.StructArray | None = None
+        self._struct_scalar: pa.StructScalar | None = None
+        if isinstance(data, pa.ChunkedArray):
+            if data.num_chunks == 0:
+                data = pa.array([], type=OME_ARROW_STRUCT)
+            else:
+                data = data.chunk(0) if data.num_chunks == 1 else data.combine_chunks()
```
Comment (MikeLippincott) [inline src/ome_arrow/tensor.py]: Should chunks be combined right away? What if we kept the array chunked for processing??

File: src/ome_arrow/tensor.py [hunk 7]
```diff
@@ -0,0 +1,877 @@
...
+                lengths = chunks_arr.value_lengths()
+                first_len = lengths[0]
+                result = bool(first_len.is_valid and int(first_len.as_py()) > 0)
+        elif self._struct_scalar is not None:
+            chunks_scalar = self._struct_scalar["chunks"]
+            if chunks_scalar.is_valid:
+                result = len(chunks_scalar.values) > 0
+        return result
+
+    def _chunk_grid(self) -> dict[str, Any]:
```
Comment (MikeLippincott) [inline src/ome_arrow/tensor.py]: even though a lot of these class methods are internal - docstrings might be nice. If not for humans but for the machines

File: src/ome_arrow/core.py [hunk 8]
```diff
@@ -496,6 +506,51 @@ def view(
...
+
+        return TensorView(
+            self._struct_array if self._struct_array is not None else self.data,
+            t=t,
+            z=z,
+            c=c,
+            roi=roi,
+            tile=tile,
+            layout=layout,
+            dtype=dtype,
```
[the remark this answers is not in the document]
Comment (d33bs) [reply src/ome_arrow/core.py]: Thanks, agreed!

File: src/ome_arrow/tensor.py [hunk 9]
```diff
@@ -0,0 +1,877 @@
...
+from typing import Any, Iterator, List, Sequence
+
+import numpy as np
+import pyarrow as pa
+import pyarrow.compute as pc
+
+from ome_arrow.export import plane_from_chunks
+from ome_arrow.meta import OME_ARROW_STRUCT
+
+_TZCHW = "TZCHW"
```
[the remark this answers is not in the document]
Comment (d33bs) [reply src/ome_arrow/tensor.py]: Order doesn't necessarily matter here but we are opinionated about what's "canon/default." The user can configure how they'd like, this is just a default. I'll add a comment or two to explain this in the code too.

File: src/ome_arrow/tensor.py [hunk 10]
```diff
@@ -0,0 +1,877 @@
...
+
+    def __init__(
+        self,
+        data: dict[str, Any] | pa.StructScalar | pa.StructArray | pa.ChunkedArray,
+        *,
+        t: int | slice | Sequence[int] | None = None,
+        z: int | slice | Sequence[int] | None = None,
+        c: int | slice | Sequence[int] | None = None,
+        roi: tuple[int, int, int, int] | None = None,
+        tile: tuple[int, int] | None = None,
```
[the remark this answers is not in the document]
Comment (d33bs) [reply src/ome_arrow/tensor.py]: Thanks good call - I added in a 3d tiling option here. 4d and 5d could get complex!

File: src/ome_arrow/tensor.py (hunk 6 above)
Comment (d33bs) [reply src/ome_arrow/tensor.py]: Nice idea! I added a `chunk_policy` with some optionality here to chunk or not based on user preference.
